br/pkg/lightning/backend/external/iter.go

```diff
@@ -292,23 +313,121 @@ func (i *mergeIter[T, R]) next() bool {
 			}
 			i.readers[i.lastReaderIdx] = nil
 			delete(i.hotspotMap, i.lastReaderIdx)
+			closeReaderIdx = i.lastReaderIdx
 		default:
 			i.logger.Error("failed to read next element",
 				zap.String("path", rd.path()),
 				zap.Error(err))
 			i.err = err
-			return false
+			return false, closeReaderIdx
 		}
 	}
 	i.lastReaderIdx = -1
 
 	if i.h.Len() == 0 {
-		return false
+		return false, closeReaderIdx
 	}
 	currMergeElem := heap.Pop(&i.h).(mergeHeapElem[T])
 	i.curr = currMergeElem.elem
 	i.lastReaderIdx = currMergeElem.readerIdx
-	return true
+	return true, closeReaderIdx
+}
+
+// limitSizeMergeIter acts like a mergeIter, except that each reader has a weight
+// and it will try to open more readers with the total weight doesn't exceed the
+// limit.
+//
+// Because it's like a mergeIter it's expected to iterate in ascending order,
+// caller should set a proper limit to do not block opening new readers
+// containing the next minimum element.
+type limitSizeMergeIter[T heapElem, R sortedReader[T]] struct {
```

Comment:
if the files are ordered by firstKey, we can open those files when current iterated key >= firstKey, so no need to calculate weight.

it's error-prone to let caller make sure the limit is enough and might cause the iterated key might not in ascending order.

Reply:
We need to open the file to know the first key (or record first key in another place). In current design, we don't record first key for intra-`MultipleFilesStat`. Another point is it's more cheap to saving a number representing the maximum overlapping level than saving every start key.

> make sure the limit is enough and might cause the iterated key might not in ascending order.

the caller can use `mergeIter` directly, like `MergeKVIter`. Currently for this problem I can only figure out `limitSizeMergeIter` with this prerequisite.

Reply:
And in fact in my first design half a year ago I suggest to record `MinKey`, `MaxKey` for every file and save it in a `MetadataStorage` interface (a table in downstream database should be OK), so these actions are easy to implement 😂 Current global sort system component design is complicated and I think they are over-optimized.